### PR TITLE
Make home page recents great again. (Rebased)

### PR DESCRIPTION
**Pull request by SiteEffect**
## Description
- Make it more clear if a Pokémon is still active
- Provide time information to recents
- Provide Google Maps to recents for a faster hunt
- Show Pokémon-IV if available (optional Feature with ability to toggle)

## Motivation and Context
See Feature Request #101 

## How Has This Been Tested?

On my webserver. 

## Screenshots
Mythics off, IVs off
bildschirmfoto 2017-01-11 um 00 40 31

Mythics off, IVs on
bildschirmfoto 2017-01-11 um 00 40 06

Mythics on, IVs on
bildschirmfoto 2017-01-11 um 00 40 20

(no chance to get time-out data for Mythics off, but please speak to my region to spawn more Mythics to also provide you with a screenshot of live and time-out Pokémon mixed)

~~Also affects pokemon.page.php in a good way~~ (already merged, now)

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## WIP - Issues and To-Do
Still buggy as hell with current master.
- [x] [Issue] Why is the newest Pokémon pushed again and again any other second
- [x] [Issue] Why get IV-bars instant-hidden
- [x] [Issue] Why get, additionally, all IV-bars hidden when aru.php pushes a new Pokémon
- [x] [Issue] Unknown issue with new functions.php by @Obihoernchen 
- [x] Auto-update time information with aru
- [x] Don't ignore people who do not scan for IVs at all, make IVs in recents optional. (done by @Obihoernchen , thanks!)
- [ ] [Issue] Lost recents between update intervals
- [x] [Issue] Bad auto-rearrange on mobile when unequal height of columns. (see screenshot in the comments)
- [ ] Am I sane? @Obihoernchen is!

### CODE REVIEW

**Comment by SiteEffect** [comment]
For old comments refer to #122

File: core/process/data.loader.php [hunk 1]
```diff
@@ -432,21 +432,33 @@
...
-	if ($result->num_rows > 0) {
-		while ($data = $result->fetch_object()) {
-			$recents[] = $data->pokemon_id;
-		}
-	}
+    if ($result->num_rows > 0) {
+        while ($data = $result->fetch_object()) {
+            $recent = new stdClass();
+            $recent->id = $data->pokemon_id;
+            $recent->last_seen = strtotime($data->disappear_time)+60*60;
```
Comment (SiteEffect) [inline core/process/data.loader.php]: @Obihoernchen I tried to remove the ``60*60`` and implement in the scheme of the pokemon.page.php request with
```
$req 		= "SELECT DISTINCT pokemon_id, (disappear_time ".$time->symbol." INTERVAL ".$time->delay." HOUR) as disappear_time, latitude, ...
```
to obtain the correct time information. This slowed down any http requests tremendously for unknown reason with my already medium-big DB.

**Comment by SiteEffect** [comment]
Latest (^this) PR-commit also addresses that ``aru.php`` now only writes the variables when new most recent one is found (except ID which needs to be written right away because of the check itself)

**Comment by Obihoernchen** [comment]
Have a look at my branch: https://github.com/Obihoernchen/Worldopole/tree/recents
696478a65fd5fa7c74d5e0f413d0d4a8e6b1d70b
0eb85515c75835904dd645088fb7f08cc842035b
3878dd6cce7c6ece7791096c4b40abc028e7ba20

**Review by SiteEffect — COMMENTED** [review]
#134 should address this one.

File: core/process/data.loader.php (hunk 1 above)
Comment (Obihoernchen) [reply core/process/data.loader.php]: Yep now it's like this: https://github.com/brusselopole/Worldopole/commit/3878dd6cce7c6ece7791096c4b40abc028e7ba20
There I use the originl row to do sorting which will allow mysql to use indexes :) 
I tested it and it's even faster than the php implementation.

**Comment by SiteEffect** [comment]
Might add #136 only if merged.

**Comment by SiteEffect** [comment]
**[Issue]**
When more then one Pokémon is "new" between the update-interval of 5 seconds, the list is not 100% true because only the very recent is pushed, while the others get lost until manually reloading of the page.

**Comment by SiteEffect** [comment]
^huge thanks to @PoUpA for finding the fix for the hidden IV-bars on auto-update.

**To-Do left:** 
(if you like to add more please feedback)
- [x] Auto-update time information with aru
- [ ] [Issue] Lost recents between update intervals
- [x] [Issue] Bad auto-rearrange on mobile when unequal height of columns. Screenshot:
bildschirmfoto 2017-01-09 um 07 35 26

**Comment by SiteEffect** [comment]
^why failed Scrutinizer?

**Comment by Obihoernchen** [comment]
Scrunzizier seems to be broken right now. Don't care about it.

**Comment by SiteEffect** [comment]
In this case, I guess this PR is ready. The open issues/to-do are probably better addressed later when aru and data.loader is reworked - those are non-breaking issues.

File: core/js/home.script.js [hunk 2]
```diff
@@ -75,3 +81,34 @@ function updateCounter(new_value, classname)
 		}
 	});
 })();
+
+// Array with timer IDs
+var timers = [];
```
Comment (Obihoernchen) [inline core/js/home.script.js:86]: Array which holds all running timer IDs

File: core/js/home.script.js [hunk 3]
```diff
@@ -58,13 +58,19 @@ function updateCounter(new_value, classname)
 		url: 'core/process/aru.php?type=spawnlist_update&last_id='+last_id,
 		success: function (data) {
 			
-			if (data != '') {
-				//console.log(data);
-				$('.last-mon-js').prepend(data);
+			if (!$.isEmptyObject(data)) {
```
Comment (Obihoernchen) [inline core/js/home.script.js:61]: Will get a json object now with index 0 = html code, 1 = timer countdown 2 = timer dom ID

File: core/js/home.script.js [hunk 4]
```diff
@@ -75,3 +81,34 @@ function updateCounter(new_value, classname)
...
+		if (--countdown >= 0) {
+			$(element).text(output);
+		} else {
+			$(element).text("- " + output);
+			$(element).css({ 'color': 'rgb(200, 50, 50)'});
+		}
+	}, 1000));
+}
+
+function stopTimer()
```
Comment (Obihoernchen) [inline core/js/home.script.js:111]: clears the interval (stops the timer) of the last element on the page + removes this element from the timer array

File: core/json/variables.examples.json [hunk 5]
```diff
@@ -25,6 +25,7 @@
 		"timezone"		:	"Europe/Paris",
 		"max_pokemon"		:	151,
 		"mythic_recents"        :       false,
+		"recents_show_iv"	:	false,
```
Comment (Obihoernchen) [inline core/json/variables.examples.json:28]: new option to toggle iv display on/off

File: core/process/aru.php [hunk 6]
```diff
@@ -146,40 +146,83 @@
...
+			$html .= '
+			</div>';
+
+			$new_spawn[] = $html;
+			$countdown = $last_seen - time();
+			$new_spawn[] = $countdown;
+			$new_spawn[] = "#timer-".time()."-js";
+
+			header('Content-Type: application/json');
+			echo json_encode($new_spawn);
```
Comment (Obihoernchen) [inline core/process/aru.php:220]: build json for core/js/home.script.js

File: core/process/aru.php [hunk 7]
```diff
@@ -146,40 +146,83 @@
...
+					</div>';
+				}
+			}
+			$html .= '
+			</div>';
+
+			$new_spawn[] = $html;
+			$countdown = $last_seen - time();
+			$new_spawn[] = $countdown;
+			$new_spawn[] = "#timer-".time()."-js";
```
Comment (Obihoernchen) [inline core/process/aru.php:217]: use current time for id. Should be unique because update happens only once every 5s.

File: pages/home.page.php [hunk 8]
```diff
@@ -48,38 +48,56 @@
...
-				<p class="pkmn-name"><a href="pokemon/<?= $pokemon ?>"><?= $pokemons->pokemon->$pokemon->name ?></a></p>
-			
+		<?php
+		foreach ($recents as $key => $pokemon) {
+			$id = $pokemon->id; ?>
+			<div class="col-md-1 col-xs-3 pokemon-single" data-pokeid="<?= $id ?>">
+				<a href="pokemon/<?= $id ?>"><img src="core/pokemons/<?= $id ?>.png" alt="<?= $pokemons->pokemon->$id->name ?>" class="img-responsive"></a>
+				<a href="pokemon/<?= $id ?>"><p class="pkmn-name"><?= $pokemons->pokemon->$id->name ?></p></a>
+				<a href="https://maps.google.com/?q=<?= $pokemon->last_location->latitude ?>,<?= $pokemon->last_location->longitude ?>&ll=<?= $pokemon->last_location->latitude ?>,<?= $pokemon->last_location->longitude ?>&z=16" target="_blank">
+					<small id="timer-<?= $key ?>-js">00:00:00</small>
```
Comment (Obihoernchen) [inline pages/home.page.php:62]: For initial 12 timers use index as ID. Should be unique as well.

File: pages/home.page.php [hunk 9]
```diff
@@ -48,38 +48,56 @@
...
+				<?php
+					}
+				} ?>
 			</div>
-			
-		<?php }?>
-		
+		<?php
+			// Array with ids and countdowns to start at the end of this file
+			$timers["#timer-".$key."-js"] = $pokemon->last_seen - time();
```
Comment (Obihoernchen) [inline pages/home.page.php:91]: I don't know whether this is a good idea to do it like this? (see below as well)

File: pages/home.page.php [hunk 10]
```diff
@@ -90,6 +108,15 @@
...
-	<?php }?>
-			
+	<?php } ?>
 </div>
+
+<script>
+	document.addEventListener('DOMContentLoaded', function() {
+		<?php
+		foreach (array_reverse($timers) as $id => $countdown) { ?>
+			startTimer(<?= $countdown ?>,"<?= $id ?>");
```
Comment (Obihoernchen) [inline pages/home.page.php:120]: This will start all 12 initial timers from above. Is it ok to do it like this?

**Review by Obihoernchen — APPROVED** [review]

**Comment by SiteEffect** [comment]
What is this Xcode indent -.-

**Comment by SiteEffect** [comment]
@PoUpA please review @Obihoernchen 's PHP questions - I can only trust with the back-end (noob is me, myself and I). Front-end seems to be merge-able, I guess.

**Comment by Obihoernchen** [comment]
merged #140

**Pull request closed without merging**